### PR TITLE
Blake2b implementation

### PR DESCRIPTION
**Pull request by ultimaterex**
Implements  Blake2b from the PointyCastle library in 256 and 512 flavours.

Adds tests for TSDK-441-implement-unit-test-cases-for-hash-blake-2-b

### CODE REVIEW

File: .idea/BramblDart.iml
Comment (KalervoHyyppa) [inline .idea/BramblDart.iml:1]: Are all of these `.idea/` files necessary? Of can they be added to the `.gitignore?`

**Comment by ultimaterex** [comment]
> See comment about `.idea/`

Interestingly I've added these to the gitignore, but it's not picking up the changes

File: lib/src/utils/extensions.dart [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import 'dart:convert';
+import 'dart:typed_data';
+
+import 'package:convert/convert.dart';
+
+extension StringExtension on String {
+  Uint8List toUint8List() {
```
Comment (SeanCheatham) [inline lib/src/utils/extensions.dart:7]: Would it help to specify in the method name that this is returning UTF-8 encoded data?  i.e. `toUtf8Uint8List`?  It looks a bit clunky, but it avoids the confusion of base58/base64/hex/utf8 in the result

File: test/crypto/hash/blake2b_test.dart [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        final hash = _doHashCheck("", Blake2b512());
+        expect(
+            hash,
+            equals(
+                "786a02f742015903c6c6fd852552d272912f4740e15847618a86e217f71f5419d25e1031afee585313896444934eb04b903a685b1448b755d56f701afe9be2ce"));
+      });
+
+      test('throws error when bytes is empty', () {
+        final blake2b = Blake2b512();
+        expect(() => blake2b.hash([]), throwsA(isA<AssertionError>()));
```
Comment (SeanCheatham) [inline test/crypto/hash/blake2b_test.dart:77]: I think an empty array is still valid (similar to the empty string case above, and should have the same output)

File: .idea/BramblDart.iml [hunk 3]
```diff
@@ -1,13 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (SeanCheatham) [inline .idea/BramblDart.iml:1]: I think you'll need to `git rm` these files in order for the gitignore to start working again.

File: test/crypto/hash/blake2b_test.dart (hunk 2 above)
Comment (ultimaterex) [reply test/crypto/hash/blake2b_test.dart:77]: Yep, had to retroactively remove these.

File: .idea/BramblDart.iml (hunk 3 above)
Comment (ultimaterex) [reply .idea/BramblDart.iml:1]: Aha, I had to retroactively remove these

File: lib/src/utils/extensions.dart (hunk 1 above)
Comment (ultimaterex) [reply lib/src/utils/extensions.dart:7]: Clunky, but does clear up the confusion. I'll change it to that

File: example/brambl_dart_example.dart
Comment (DiademShoukralla) [inline example/brambl_dart_example.dart:1]: For my own understanding, what is the purpose of this file?

**Review by SeanCheatham — APPROVED** [review]

File: test/crypto/hash/blake2b_test.dart [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+      });
+
+      test('hash empty list', () {
+        final hash = Blake2b256().hash([]).toHexString();
+        expect(hash, equals("0e5751c026e543b2e8ab2eb06099daa1d1e5df47778f7787faab45cdf12fe3a8"));
+      });
+
+      test('throws error when bytes is empty', () {
+        final blake2b = Blake2b256();
+        // expect(() => blake2b.hash([]), throwsA(isA<AssertionError>()));
```
Comment (DiademShoukralla) [inline test/crypto/hash/blake2b_test.dart:43]: (also line 47) is there a reason that these commented out lines are checked in?

**Review by DiademShoukralla — APPROVED** [review]
Looks good! Just 2 minor comments/questions

File: example/brambl_dart_example.dart
Comment (ultimaterex) [reply example/brambl_dart_example.dart:1]: This is standard in pub packages where you show an "example" implementation class
eg: https://pub.dev/packages/html/example

File: test/crypto/hash/blake2b_test.dart (hunk 4 above)
Comment (ultimaterex) [reply test/crypto/hash/blake2b_test.dart:43]: good catch! fixed in latest

**Pull request merged**